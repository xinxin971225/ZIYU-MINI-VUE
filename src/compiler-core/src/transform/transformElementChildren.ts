import { NodeTypes } from "../ast";
import { isTextOrInterpolation } from "../utils";

export function transformElementChildren(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const children = node.children;
      // 对原有元素children 除了element外的其他节点做转换-》添加一个+号 -》收集为一个新的节点类型
      let currentChild;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isTextOrInterpolation(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const nextChild = children[j];
            if (isTextOrInterpolation(child)) {
              if (!currentChild) {
                currentChild = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child],
                };
              }
              currentChild.children.push(" + ");
              currentChild.children.push(nextChild);
              // 删除元素时，数组的内容会移位，所以j要保持上一次的位置
              children.splice(j, 1);
              j--;
            } else {
              currentChild = null;
            }
          }
        }
      }
    };
  }
}
