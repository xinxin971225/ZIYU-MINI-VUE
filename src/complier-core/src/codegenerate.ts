import { isString } from "../../share";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_BLOCK,
  helpersNameMap,
  OPEN_BLOCK,
  TO_DISPLAY_STRING,
} from "./runtimeHelper";

export function generate(ast) {
  const context = createGenerateContext();
  const { push } = context;
  getFunctionPreamble(context, ast);
  push("export ");
  getFunctionNameAndArgs(context);

  genNode(context, ast.genCodeNode);
  push(" }");
  return {
    code: context.code,
  };
}

function getFunctionPreamble(context, ast) {
  const { push } = context;
  const VueBinging = "vue";
  const aliasHelper = (s) => `${s} as _${s}`;
  if (ast.helpers.length) {
    push(
      `import { ${ast.helpers
        .map(aliasHelper)
        .join(", ")} } from "${VueBinging}"`
    );
  }
  push("\n");
}

function createGenerateContext() {
  const context: any = {
    code: "",
    push: (str) => {
      context.code += str;
    },
    getHelperName: (key) => {
      return helpersNameMap[key];
    },
  };
  return context;
}

function getFunctionNameAndArgs(context) {
  const { push } = context;
  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const argStr = args.join(", ");
  push(`function ${functionName}(${argStr}){`);
  push(" return ");
}
function genNode(context, node) {
  switch (node.type) {
    case NodeTypes.TEXT:
      getTextCode(context, node);
      break;
    case NodeTypes.INTERPOLATION:
      getInterpolation(context, node);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      getExpression(context, node);
      break;
    case NodeTypes.ELEMENT:
      getElement(context, node);
      break;
    case NodeTypes.COMPOUND_EXPRESSION:
      getCompoundExpression(context, node);
      break;
    default:
      break;
  }
}

function getCompoundExpression(context, node) {
  const { push } = context;
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      push(child);
    } else {
      genNode(context, child);
    }
  }
}
function getElement(context, node) {
  const { push, getHelperName } = context;
  const { tag, props, genCodeNode } = node;
  push(
    `(_${getHelperName(OPEN_BLOCK)}(), _${getHelperName(
      CREATE_ELEMENT_BLOCK
    )}( `
  );
  const fixValList = fixUndefinedVal([tag, props, genCodeNode]);
  genNodeList(fixValList, context);
  push(", 1 /* TEXT */))");
}
function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(`${node}, `);
    } else {
      genNode(context, node);
    }
  }
}
function fixUndefinedVal(vals) {
  return vals.map((v) => v || "null");
}

function getTextCode(context, node) {
  const { push } = context;

  const text = node.content;
  push(`"${text}"`);
}
function getInterpolation(context: any, node: any) {
  const { push, getHelperName } = context;
  const rawContent = node.content;
  push(`_${getHelperName(TO_DISPLAY_STRING)}(`);
  genNode(context, rawContent);
  push(")");
}
function getExpression(context: any, node: any) {
  const { push } = context;
  // 这里的ctx属于半业务了，应该交给transform 这里只负责拼接字符串
  // push(`_ctx.${node.content}`);
  push(`${node.content}`);
}
