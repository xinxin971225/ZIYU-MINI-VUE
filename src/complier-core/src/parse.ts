import { NodeTypes } from "./ast";

export function baseParse(content: string) {
  const context = createParseContext(content);
  return createRoot(parseChildren(context));
}

function parseChildren(context) {
  const nodes: any[] = [];
  const s = context.source;
  if (s.startsWith("{{")) {
    const node = parseInterpolation(context);
    nodes.push(node);
  } else if (/^<[a-z]/i.test(s)) {
    const node = parseElement(context);
    nodes.push(node);
  }

  return nodes;
}
function parseElement(context) {
  const element = parseTag(context);
  parseTag(context);

  return element;
}
function parseTag(context) {
  // 解析出div
  const match: any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];
  // 删掉解析过的东西
  advanceBy(context, match[0].length + 1);
  return {
    type: NodeTypes.ELEMENT,
    tag: tag,
  };
}
function parseInterpolation(context) {
  // source -> "{{message}}"
  const startDelimiter = "{{";
  const closeDelimiter = "}}";
  const { source } = context;
  const closeIndex = source.indexOf(closeDelimiter, startDelimiter.length);
  const rawContent = source.slice(startDelimiter.length, closeIndex);
  const content = rawContent.trim();
  advanceBy(context, closeIndex + closeDelimiter.length);
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  };
}
function advanceBy(context, length) {
  context.source = context.source.slice(length);
}

function createRoot(children) {
  return {
    children,
  };
}

function createParseContext(content: string) {
  return {
    source: content,
  };
}
