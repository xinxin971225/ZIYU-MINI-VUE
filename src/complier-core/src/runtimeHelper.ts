export const TO_DISPLAY_STRING = Symbol("toDisplayString");
export const OPEN_BLOCK = Symbol("openBlock");
export const CREATE_ELEMENT_BLOCK = Symbol("createElementBlock");
export const helpersNameMap = {
  [TO_DISPLAY_STRING]: "toDisplayString",
  [OPEN_BLOCK]: "openBlock",
  [CREATE_ELEMENT_BLOCK]: "createElementBlock",    
};
