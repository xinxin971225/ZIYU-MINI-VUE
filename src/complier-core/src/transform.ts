export function transform(root, options = {}) {
  const context = createRootContext(root, options);
  traverseNode(root, context);
  createGenCode(root);
}

function createGenCode(root) {
  root.genCodeNode = root.children[0];
}

/**
 * 创建一个全局上下文对象
 * @param root
 * @param options
 * @returns
 */
function createRootContext(root, options) {
  return {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };
}

/**
 * 处理每个节点的方法
 * @param node
 */
function traverseNode(node, context) {
  // // 这块逻辑不应该写在for循环中
  // if (node.type === NodeTypes.TEXT) {
  //   node.content += "ziyu";
  // }
  const { nodeTransforms } = context;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transformFn = nodeTransforms[i];
    transformFn(node);
  }
  traverseNodeChildren(node, context);
}

function traverseNodeChildren(node, context) {
  // 相对固定的深度优先遍历
  const children = node.children;
  if (children) {
    // 深度优先遍历
    for (let i = 0; i < children.length; i++) {
      const nodeChildren = children[i];
      traverseNode(nodeChildren, context);
    }
  }
}
