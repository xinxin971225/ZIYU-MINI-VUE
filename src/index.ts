// mini-vue 入口
export * from "./runtime-core/index";
export { h } from "./runtime-core/h";
export { renderSlots } from "./runtime-core/helper/renderSlots";
