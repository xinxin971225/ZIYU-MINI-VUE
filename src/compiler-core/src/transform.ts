import { NodeTypes } from "./ast";
import { helpersNameMap, TO_DISPLAY_STRING } from "./runtimeHelper";

export function transform(root, options = {}) {
  const context = createRootContext(root, options);
  traverseNode(root, context);
  createGenCode(root);
  createHelpers(root, context);
}

function createHelpers(root, context) {
  root.helpers = [...context.helpers.keys()];
}

function createGenCode(root) {
  const children = root.children[0];
  if (children.type === NodeTypes.ELEMENT) {
  }
  root.genCodeNode = children;
}

/**
 * 创建一个全局上下文对象
 * @param root
 * @param options
 * @returns
 */
function createRootContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    addHelper: (key) => {
      context.helpers.set(key, 1);
    },
    getHelperName: (key) => {
      return helpersNameMap[key];
    },
  };
  return context;
}

/**
 * 处理每个节点的方法
 * @param node
 */
function traverseNode(node, context) {
  // // 这块逻辑不应该写在for循环中
  // if (node.type === NodeTypes.TEXT) {
  //   node.content += "ziyu";
  // }
  const onExitFn: any = [];
  const { nodeTransforms, addHelper, getHelperName } = context;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transformFn = nodeTransforms[i];
    // 对内部节点的破坏性修改，应该在最后执行
    const onExit = transformFn(node, context);
    if (onExit) {
      onExitFn.push(onExit);
    }
  }
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      addHelper(getHelperName(TO_DISPLAY_STRING));
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseNodeChildren(node, context);
      break;
    default:
      break;
  }
  let i = onExitFn.length;
  while (i--) {
    onExitFn[i]();
  }
}

function traverseNodeChildren(node, context) {
  // 相对固定的深度优先遍历
  const children = node.children;
  if (children) {
    // 深度优先遍历
    for (let i = 0; i < children.length; i++) {
      const nodeChildren = children[i];
      traverseNode(nodeChildren, context);
    }
  }
}
