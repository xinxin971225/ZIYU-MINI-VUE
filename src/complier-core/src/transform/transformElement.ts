import { NodeTypes } from "../ast";
import { CREATE_ELEMENT_BLOCK, OPEN_BLOCK } from "../runtimeHelper";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    const { getHelperName } = context;
    context.addHelper(getHelperName(OPEN_BLOCK));
    context.addHelper(getHelperName(CREATE_ELEMENT_BLOCK));
  }
}
