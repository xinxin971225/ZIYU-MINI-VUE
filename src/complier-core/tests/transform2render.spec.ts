import { generate } from "../src/codegenerate";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transform/transformElement";
import { transformElementChildren } from "../src/transform/transformElementChildren";
import { transformExpression } from "../src/transform/transformExpression";

describe("ast -> render ", () => {
  it("happy path simple str", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
  it("happy path simple interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("happy path simple element", () => {
    const ast = baseParse("<div>hi,{{message}}</div>");
    transform(ast, {
      nodeTransforms: [
        transformExpression,
        transformElement,
        transformElementChildren,
      ],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
