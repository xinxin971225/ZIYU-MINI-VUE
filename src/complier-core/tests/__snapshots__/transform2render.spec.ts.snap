// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast -> render  happy path simple element 1`] = `
"import { openBlock as _openBlock, createElementBlock as _createElementBlock, toDisplayString as _toDisplayString } from \\"vue\\"
export function render(_ctx, _cache){ return (_openBlock(), _createElementBlock(\\"div\\", null \\"hi,\\" +  _toDisplayString(_ctx.message), 1 /* TEXT */)) }"
`;

exports[`ast -> render  happy path simple interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from \\"vue\\"
export function render(_ctx, _cache){ return _toDisplayString(_ctx.message) }"
`;

exports[`ast -> render  happy path simple str 1`] = `
"
export function render(_ctx, _cache){ return \\"hi\\" }"
`;
